### DISCUSSION TITLE
Site redesign

### DISCUSSION THREAD

**Discussion by JeffWScott** [Lamden/docs#20] [General]
📢  To all the developers/potential developers!  

We are going to focus some energy on getting https://docs.lamden.io retooled and redesigned to make it more relevant for developers.  

To avoid confusion https://developers.lamden.io/ was not created by Lamden and the maintainer of that has not been around the community in a long time so we can't make changes to it.  https://docs.lamden.io is a simpler site and has more current documentation.

To get the discussion started I have compiled a list of questions that we would love answers to!
❓ What is the site missing?
❓ What would make the site easier to use?
❓ What is confusing about the site
❓ When you think about developing a dapp, what is the first thing you look for and why doesn't the site give you that?
❓ What doe the site do well?
❓ What code examples do you think would be beneficial?
❓  What "getting started" steps do you wish were there?
❓ What parts of https://developers.lamden.io do you think work and we should copy over to the other site?

Also, any other comments or feedback are welcome!

**Comment by Endogen** [comment]
❓ What is the site missing?
I'm thinking about this as a developer portal but maybe that's a bit too much and should be it's own thing and not necessarily integrated

1) Code examples would be great. A dev should be able to check out on some kind of example how to code up a contract and integrate it into a website.

2) Then a repository of existing handpicked contracts would be great. Then new devs could check out what already exists and how it has been realized or could directly start using it. It would be great if we could expand this into more than just a list of contracts. Something like a platform where every contract has a maintainer and a version and users can comment on certain parts of the code so that contracts "develop" over time and mature and are built in a generic way to allow for reuse.

3) Coding pattern like Action Core and best practises

4) Integrating bounties and developments that we'd love to get done so that new devs can pick it up

❓ What would make the site easier to use?
Maybe some kind of guiding system. Something like this: https://www.djangoproject.com/start/

❓ What is confusing about the site
I can imagine that if you're new, you don't know where to start. Do i need to access the nodes REST interface to interact with the chain? Is there an IDE or can i just code in PyCharm? If i want to build a good dev environment to build stuff, what do i need (tak the content from your medium post about setting up the dev server)?

❓ When you think about developing a dapp, what is the first thing you look for and why doesn't the site give you that?
How would i start doing what i want? Maybe we could have an interactive workflow that guides you through what's needed:
--> Write smart contract 
        --> will it be a desktop app? --> how to integrate Lamden into app
        --> will it be a web app? --> how to interact with Lamden via JS and show results
       ---> if it's a token, which fatures could it have? Linking to LST001 & 2

Or having specific examples shown:
Assume you want to integrate Lamden with some supply chain stuff. Why contracts would you need? Write contract, write website for tracking the thing based on GPS with access to datapoints that are on chain to verify state. Then client apps to interact with the chain / contract to add datapoints so that they show up on the website etc.

❓ What code examples do you think would be beneficial?
- Token contract
- Saving and retrieving text messages
- Locking tokens (showing how datetime stuff works)
- Generating random values and hashes and how they could be used
- Saving game states

❓ What "getting started" steps do you wish were there?
I think there should be the possibility to directly dive hardcore in to anything that experienced devs need, like documentation about contracting and REST API and wallet integration and GitHub repos but also a path for new devs that need guidance and can click through some kind of tutorial to check things out.

❓ What parts of https://developers.lamden.io/ do you think work and we should copy over to the other site?
I like the "Meet the Lamdne Toolset" overview where you can directly dive into what you need if you know what you need.

**Comment by tautraders** [comment]
Not a developer, but some things that might be helpful:

- "Search" box so users can use keywords to search for specific documentation
- Link to the standards documentation (https://github.com/Lamden-Standards) to promote the use of standards across the ecosystem
- Newbie guides for walking new developers through the steps to launch a basic token or dapp from scratch (similar to your Medium article series but I think it would be helpful to have a menu directory/steps sequenced out - https://blog.lamden.io/smart-contracting-with-python-2af233620dca)

**Comment by Dapiguabc** [comment]
❓ What code examples do you think would be beneficial?

An example of dapp demonstrates a wide range of  Lamden functionality. Including following aspects:
1.  How to write a  complex contract
2.  Show how to interact with wallet, contract and block service
3.  Detailed code annotation

**Reply by JeffWScott** [reply]
This is some great feedback!

How does everyone feel about this approach?

Utilize the current docs site, but change these pages to have both "getting started" and more advanced topics. 
https://docs.lamden.io/develop
https://docs.lamden.io/wallet
https://docs.lamden.io/smart_contracting

We can have links to the actual docs as need be.

Examples of what to add for https://docs.lamden.io/smart_contracting

1) Code example needs to be changed to the LST001 standard contract.

2) Add steps to upload the contract to the blockchain

3) Add steps to access data from the contract using the block service (we should make the testnet block service accessible to developers)

4) Add some more advanced contract examples

5) Add Links to the Lamden Standards contracts

**Reply by JeffWScott** [reply]
![Uploading captures_chrome-capture-2022-5-10.png…]()

What do we think of this approach?? Is this better? We could do something like this for each product.

**Comment by JeffWScott** [comment]
Are there any contracting ideas that you want better examples or examinations of?

**Reply by Dapiguabc** [reply]
> Are there any contracting ideas that you want better examples or examinations of?

Actually, I prefer something like this(https://ethereum.org/en/developers/):
![U~CMGONO58KP@H(F2)_AOHK](https://user-images.githubusercontent.com/17672485/172940950-e9d563e9-0b36-4162-95dc-d75368165731.png)

**Reply by JeffWScott** [reply]
This a good model for sure.  Can we replicate this with our current documentation? Or do we need to create some of these steps?

**Reply by Endogen** [reply]
Code snippets for contract functions:
- How to execution a function only at specific datetime or after specific time passed (example could be time locking tokens)
- How to get current price of a token from rocketswap DEX
- How to store message in contract and building a simple messanger (https://github.com/Endogen/gammaphi-contracts/blob/master/core/con_gamma_phi_messenger_v1.py)

Not contracting related:
- How to create a transaction and send TAU around (whole interaction explained with requesting nonce and providing arguments, using the `build_transaction()`, etc.)
- How to submit a contract
- How to write a simple test for your contract to test locally

Most of that is already part of the contracting docs but might be good to have it explicitly showed and explained with code snippets.

It could also be good to explain some basic concepts:
- Low level stuff like how to generate addresses (https://github.com/Endogen/lamden-python-dev/blob/master/lamden-address.py)
- How are addresses technically generated? 64 chars long, any random 32 bytes are a valid private or public key
- How to generate a wallet (https://github.com/Endogen/lamden-python-dev/blob/master/lamden-wallet.py)
- How to check if a transaction was successful (https://github.com/Endogen/lamden-python-dev/blob/master/lamden-check-tx.py) or maybe doing that via block service

**Reply by Dapiguabc** [reply]
> This a good model for sure. Can we replicate this with our current documentation? Or do we need to create some of these steps?

Of course we can. Just need to add these step cards to home page and have these cards linked to the actual docs.

**Comment by Endogen** [comment]
We could in general have an official contract repo on GitHub to which users could create PRs to add contracts to. Contracts would only be merged into the repo if they have extensive comments and are build in a general way so that they are as reusable as possible. Goal would be to let people easily copy these, adjust minimally and use for their own use cases.

That repo could be curated by some devs
